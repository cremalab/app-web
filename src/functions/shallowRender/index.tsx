import { createRenderer } from "react-test-renderer/shallow"
const renderer = createRenderer()

export const shallowRender = (x: JSX.Element) => renderer.render(x)
