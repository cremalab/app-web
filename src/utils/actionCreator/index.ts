/* eslint-disable @typescript-eslint/no-explicit-any */
import { Action } from "../../types/Action"
import { ActionType } from "../../types/ActionType"

/**
 * `actionCreator`
 * ---------------------------------------------
 * Returns a redux action creator
 */

export const actionCreator = <
  Type extends PropertyKey = ActionType,
  Args extends any[] = [],
  Payload = undefined
>(
  type: Type,
  makePayload: (...args: Args) => Payload = (...x) => x[0],
) => (...args: Args): Action<Type, Payload> => ({
  type,
  payload: makePayload(...args),
})
