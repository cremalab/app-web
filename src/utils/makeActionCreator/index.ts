/* eslint-disable @typescript-eslint/no-explicit-any */
import { Action } from "../../types/Action"
import { ActionType } from "../../types/ActionType"

/**
 * `makeActionCreator`
 * ---------------------------------------------
 * Makes a Redux "Action Creator" function
 */

export type MakePayload<Args extends any[], Payload> = (
  ...args: Args
) => Payload

export const makeActionCreator = <
  Payload = null,
  Args extends any[] = [Payload],
  Type extends ActionType = ActionType
>(
  type: Type,
  makePayload: MakePayload<Args, Payload> = (...x) => x[0],
) => (...args: Args): Action<Payload, Type> => ({
  type,
  payload: makePayload(...args),
})
