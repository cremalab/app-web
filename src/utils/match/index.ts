import { Cases } from "../../types/Cases"
import { Tagged } from "../../types/Tagged"

// tslint:disable: jsdoc-format
/**
 * match
 * ---
 * Takes a `Cases` object, and the value to match against. Returns value from each case.
 *
 * Example:
 ```
const user = None
match<Maybe<User>, string>(
  {
    [MaybeTag.Some]: x => x.data.name,
    [MaybeTag.None]: () => "This user doesn't exist" 
  },
  user
) // "This user doesn't exist"
```
 */

export function match<
  TaggedUnion extends Tagged<Tag, unknown>,
  Return,
  Tag extends PropertyKey = PropertyKey
>(cases: Cases<TaggedUnion, Return, Tag>, value: TaggedUnion): Return {
  return cases[value.tag](value.data)
}
