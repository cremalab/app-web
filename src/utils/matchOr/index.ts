import { Cases } from "../../types/Cases"
import { Tagged } from "../../types/Tagged"

// tslint:disable: jsdoc-format
/**
 * matchOr
 * ---
 * Takes a partial `Cases`, a default value, and the value to match against. Returns default value if no match is present.
 *
 * Example:
 ```
const user = None
matchOr<Maybe<User>, string>(
  {
    [MaybeTag.Some]: x => x.data.name,
    // [MaybeTag.None]: () => "This user doesn't exist" 
    // ðŸ‘† partial because `Cases` lacks `None`
  },
  "This user doesn't exist",
  user
) // "This user doesn't exist"
```
 */

export function matchOr<
  TaggedUnion extends Tagged<Tag>,
  Return,
  Tag extends PropertyKey = PropertyKey,
  CaseDefault extends Return = Return
>(
  cases: Partial<Cases<TaggedUnion, Return, Tag>>,
  caseDefault: CaseDefault,
  value: TaggedUnion,
) {
  const c = cases[value.type]
  return c ? c(value) : caseDefault
}
