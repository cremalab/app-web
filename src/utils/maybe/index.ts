import { Maybe, None, Some } from "../../types/Maybe"

/**
 * maybe
 * ---
 * Returns `Maybe<A>` from possibly Nil value (`undefined` | `null`)
 */

export const maybe = <A>(a?: A | undefined | null): Maybe<A> =>
  a === undefined || a === null ? None : Some(a)
