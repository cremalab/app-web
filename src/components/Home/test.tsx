import { Home } from "."
import React from "react"
import { shallowRender } from "../../utils/shallowRender"

describe("Home", () => {
  it("renders", () => {
    // Arrange
    const valA = Home

    // Act
    const received = shallowRender(<Home>{valA}</Home>)

    // Assert
    expect(received).toMatchSnapshot()
  })
})
