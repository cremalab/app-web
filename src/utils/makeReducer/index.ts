import { match, Cases } from "../match"
import { Tagged } from "../../types/Tagged"
import { Reducer } from "redux"

export type MakeReducerCases<S, A extends Tagged> = (
  state: S,
  stateInitial: S,
) => Cases<A, S>

export const makeReducer = <S, A extends Tagged>(
  makeReducerCases: MakeReducerCases<S, A>,
  initialState: S,
): Reducer<S, A> => (state = initialState, action) => {
  const cases = makeReducerCases(state, initialState)
  return match<A, S>(cases, () => state)(action)
}
