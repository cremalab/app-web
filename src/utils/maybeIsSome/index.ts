import { Maybe, MaybeTag } from "../../types/Maybe"
import { match } from "../match"

/**
 * maybeIsSome
 * ---
 * Takes a `Maybe` value (`A`) and returns `true` if it contains a value and `false` if it does not.
 */

export const maybeIsSome = <A>(a: Maybe<A>) =>
  match<Maybe<A>, boolean>(
    {
      [MaybeTag.Some]: () => true,
      [MaybeTag.None]: () => false,
    },
    a,
  )
