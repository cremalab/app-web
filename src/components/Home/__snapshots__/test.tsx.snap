// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders 1`] = `
<div
  className="Home"
>
  [Function]
</div>
`;
